proposals/p3763.md

```diff
@@ -0,0 +1,858 @@
+> could be achieved with syntax such as:
+>
+> ```carbon
+> base class A {
+>   virtual fn Clone[self: Self]() -> A*;
+> }
+> base class B {
+>   virtual impl fn Clone[self: Self]() -> B*;
+> }
+> ```
+>
+> Here, a call to `b->Clone()` would find `B.Clone` rather than `A.Clone`, and
+> so would have return type `B*`. The downside is that the vtable for `B` would
+> have two `Clone` slots, for `A.Clone` and `B.Clone`, whereas a covariant
+> return in C++ would only need a single vtable slot to express the same thing.
+
+It is an error for a class with a custom value representation to declare or
+implement a virtual function that passes `self` by value.
+
+### `extern` declarations
+
+For `extern` declarations, some additional considerations apply to the rule
+disallowing redeclarations from syntactically differing:
+
+-   This rule cannot in general be checked during compilation. For example,
+    there may be no file that imports both an `extern` declaration and the
+    library that owns the entity.
+-   Despite being syntactically identical, `extern` declarations can have
+    different meanings from the entity they redeclare due to unqualified lookup
+    results differing.
+-   A constraint that the declarations be syntactically identical is a greater
+    burden, because it introduces syntactic coupling across libraries that may
+    make refactoring harder.
+
+We address these points as follows:
+
+-   Checking whether `extern` declarations properly redeclare their targets
+    during compilation is best-effort.
+    -   `extern` declarations are expected to be fully checked at link time,
+        including both syntactic and semantic checks, by emitting extra
+        information into the object files.
+    -   Some level of semantic checks may be important for an implementation in
+        order to ensure its data structures are consistent. For example,
+        constant evaluation may cross between file boundaries, and it may be
+        important that types used in compile-time functions have the same
+        meaning in the caller and callee in order to support such calls.
+    -   Any further compile-time checking is optional and best-effort. Carbon
+        implementations are encouraged to perform a reasonable set of checks
+        during compilation in order to improve the quality of diagnostics and
+        produce diagnostics earlier, but can defer harder cases until link time.
+        For example, an implementation could choose to not track the syntactic
+        form of a declaration, and perform only semantic checks during
+        compilation, deferring syntactic checking until link time.
+-   To support `extern` declarations, an additional constraint is imposed: the
+    lookup results for names used in each declaration of an entity are required
+    to be the same. This rule actually applies to all declarations of all
+    entities, but it only has an effect -- and needs to be checked -- for
+    `extern` declarations.
```

Comment:
Hm. For the `api` / `impl` case, we use this rule from above:

> If an entity is declared as non-`extern` in an `api` file and redeclared in an `impl` file of the same library, it is an error for the declaration to use any unqualified name that was introduced in the `impl` file.

My intent was that this would be a lot easier to check than the rule for `extern` declarations, because we don't need to do a comparison of two different sets of lookup results, we only need to track whether name lookup found a local-only result while processing a declaration -- and the idea was that if you satisfy that rule, then you don't need to check this one (and thus we don't need to track what the lookup results were for all declarations, only for `extern` declarations).

But I think maybe I was wrong. For example:

```
package Foo library "other";
namespace N;
class N.C {}
```

```
package Foo library "problem";
namespace N;
class C {}
fn N.F(c: C);
```

```
impl package Foo library "problem";
import library "other";
fn N.F(c: C) {}
```

Here, unqualified lookup for `C` in the implementation file will find `N.C`, which wasn't declared in the same file, but the declaration of `N.F` is syntactically identical to that in "other".

So I think we have two choices:

1) Change the rule above to reject a declaration in an implementation file that redeclares something in an API file unless all the lookups found something that the API file could also see (which seems like a much less clean rule), or
2) Apply this "lookup results are the same" check everywhere.

I think we should pick whichever choice is best for the implementation. How difficult would it be to track whether a lookup finds results that would have been visible in the API file? (Eg, can we put a tag on each lookup result to say whether it could only be found due to declarations *and imports* in the current file?) Conversely, how expensive would it be to perform the "lookup results are the same" check as part of all redeclaration checks?

Reply:
That's similar to the example I had in mind. Maybe more significant than I was thinking?

Regarding (2), I think it's easy:

All parameters should get types assigned while parsing the function. The incremental expense of verifying they're the same seems small: we've already implemented the code that does it, and it's just comparing instruction types in an instruction block.

Regarding (1), I guess I'm not too sure how you intended to do the checking at present. By the proposal, you're planning to allow something like:

```
class C;
namespace NS;
fn NS.F(x: C);
fn NS.F(x: C) {}
class NC.C;
```

And disallow something like:

```
class C;
namespace NS;
fn NS.F(x: C);
fn NS.G(x: C);
fn NS.F(x: C) {}
class NC.C;
fn NS.G(x: C) {}
```

Also disallowing:

```
class C;
namespace NS;
fn NS.F(x: C);
alias NS.C = C;
fn NS.F(x: C);
```

Is that correct, that you are allowed to shadow a name so long as you don't reuse it in a redeclaration?

If shadowing that case is disallowed, you can poison name lookup. Each poison entry can be exported and imported just by the implementation. Note, I think the implementation file design doesn't say we implicitly import everything the API does: we would need to add that to make this work (to ensure that not just poison values matched, but everything inbetween matches too). Then, the error could be on the `import library "other";` that introduces a conflict with a poisoned value.

Alternately, if shadowing is allowed so long as it's not later used in a redeclaration, it isn't straight poisoning of name lookup -- that would render the first example invalid. It's not just tracking which type a name resolved to, because that wouldn't catch (3). Were you planning on, for each poisoned name, keep a vector of all declarations that used it? That same vector could be exported, and validated against the redeclaration site, although it too feels a little expensive.

Reply:
Regarding (2): For a templated function, we don't have a type assigned to a parameter until we instantiate, so comparing parameter types is not enough to ensure that lookups produce the same result in that case. Also, it would be possible for the lookup results to differ and for us to get the same type anyway (though that's a less important problem I suppose, because we can at least be confident the declarations are consistent). I think we could do a lock-step comparison of the instruction blocks and make sure that all `name_ref`s have the same constant value, though that's also not quite right, because the constant values being equal doesn't imply that the lookups found the same thing, but I think we could make that work. Another option would be to track a list of all name lookups and the corresponding lookup results on the entity for an `extern` declaration, and compare that as part of the redeclaration logic (eg, by redoing all the lookups and checking they result in the same thing).

Regarding (1): I think your first example is invalid in this proposal as it stands because it violates

> In a declarative scope, it is an error if a name is first looked up and not found, and later introduced.

for the name `N.C`, but would be valid if the final declaration were moved to an implementation file. But as you say:

> If shadowing that case is disallowed, you can poison name lookup. Each poison entry can be exported and imported just by the implementation. Note, I think the implementation file design doesn't say we implicitly import everything the API does: we would need to add that to make this work (to ensure that not just poison values matched, but everything inbetween matches too). Then, the error could be on the import library "other"; that introduces a conflict with a poisoned value.

I think I like that idea. So then, to summarize, the plan would be:

- The implementation files import not just the API file, but also everything the API file imports, and all the poisoned names. So in general, they start in the same name lookup state that the API file ended in.
- Same-package imports check against the poisoned state as they add names. (For cross-package names, there's not a problem because unqualified lookup never looks in other packages.)

Plus additional checks for `extern` declarations.

Reply:
> Regarding (1): I think your first example is invalid in this proposal as it stands because it violates
> 
> > In a declarative scope, it is an error if a name is first looked up and not found, and later introduced.

Good, I was missing this interaction.

> I think I like that idea. So then, to summarize, the plan would be:
> 
> * The implementation files import not just the API file, but also everything the API file imports, and all the poisoned names. So in general, they start in the same name lookup state that the API file ended in.
> * Same-package imports check against the poisoned state as they add names. (For cross-package names, there's not a problem because unqualified lookup never looks in other packages.)
> 
> Plus additional checks for `extern` declarations.

SGTM -- I think this will be a usability improvement for implementation files in general, beyond this.

Reply:
The changes look good to me. Feel free to mark this as resolved, leaving it open in case chandlerc has more comments.